Tests/MapboxDirectionsTests/WaypointTests.swift

```diff
@@ -103,4 +103,36 @@ class WaypointTests: XCTestCase {
         waypoint.headingAccuracy = 720
         XCTAssertEqual(waypoint.headingDescription, "90.5,180.0")
     }
+    
+    func testEquality() {
+        let left = Waypoint(coordinate: CLLocationCoordinate2D(), coordinateAccuracy: nil, name: nil)
+        XCTAssertEqual(left, left)
+        
+        var right = Waypoint(coordinate: CLLocationCoordinate2D(latitude: 1, longitude: 1), coordinateAccuracy: nil, name: nil)
+        XCTAssertNotEqual(left, right)
+        
+        right = Waypoint(coordinate: CLLocationCoordinate2D(), coordinateAccuracy: 0, name: nil)
+        XCTAssertNotEqual(left, right)
+        
+        right = Waypoint(coordinate: CLLocationCoordinate2D(), coordinateAccuracy: nil, name: "")
+        XCTAssertNotEqual(left, right)
+    }
+    
+    func testTracepointEquality() {
+        let left = Tracepoint(coordinate: CLLocationCoordinate2D(), countOfAlternatives: 0, name: nil)
+        XCTAssertEqual(left, left)
+        
+        let right = Tracepoint(coordinate: CLLocationCoordinate2D(), countOfAlternatives: 0, name: nil)
+        XCTAssertEqual(left, right)
+        
+        // FIXME: Only Waypoint.==(_:_:) ever gets called: <https://stackoverflow.com/a/28794214/4585461>. This will be moot once Tracepoint becomes a struct that doesn’t inherit from Waypoint: <https://github.com/mapbox/MapboxDirections.swift/pull/388>.
+//        right = Tracepoint(coordinate: CLLocationCoordinate2D(latitude: 1, longitude: 1), countOfAlternatives: 0, name: nil)
+//        XCTAssertNotEqual(left, right)
```

Comment:
Unfortunately, it isn’t currently possible to fully compare two Tracepoints; the comparison only considers Waypoint properties. It [isn’t possible for a subclass to override `==`](https://stackoverflow.com/a/28794214/4585461) in the same manner that we had been used to overriding `isEqual(_:)`. Even comparing `Tracepoint() == Tracepoint()` calls `Waypoint.==(_:_:)` directly without involving `Tracepoint.==(_:_:)`, as seen in the code coverage report.

Fortunately, #388 would address this design flaw by converting Waypoint and Tracepoint to structs and relating them through a protocol rather than inheritance.

Unfortunately, this problem also affects the abstract DirectionsResult class. Equality is very important to a library such as MapboxDirections.swift. This makes the case for abolishing abstract classes in favor of protocols even stronger.

Reply:
#392 tracks eliminating the DirectionsOptions and DirectionsResult abstract classes.